units/patch_intent.go

```diff
@@ -648,109 +648,182 @@ func (j *patchIntentProcessor) buildTasksAndVariants(patchDoc *patch.Patch, proj
 
-func setTasksToPreviousFailed(patchDoc, previousPatch *patch.Patch, project *model.Project) error {
-	var failedTasks []string
-	for _, vt := range previousPatch.VariantsTasks {
-		tasks, err := getPreviousFailedTasksAndDisplayTasks(project, vt, previousPatch.Version)
-		if err != nil {
-			return err
+// setToFilteredTasks sets the tasks/variants based on a previous patch filters them to activated tasks and
+// based on the failedOnly flag. It adds dependencies and task group tasks as needed.
+func setToFilteredTasks(patchDoc, reusePatch *patch.Patch, project *model.Project, failedOnly bool) error {
+	activatedTasks, err := filterToActiveForReuse(reusePatch, project)
+	if err != nil {
+		return errors.Wrap(err, "getting dependencies for activated tasks")
+	}
+
+	activatedTasksDispNames := []string{}
+	failedTaskDisplayNames := []string{}
+	failedTasks := []task.Task{}
+	for _, t := range activatedTasks {
+		activatedTasksDispNames = append(activatedTasksDispNames, t.DisplayName)
+		if failedOnly && evergreen.IsFailedTaskStatus(t.Status) {
+			failedTasks = append(failedTasks, t)
+			failedTaskDisplayNames = append(failedTaskDisplayNames, t.DisplayName)
+		}
+	}
+	filteredTasks := activatedTasksDispNames
+
+	patchDoc.Tasks = filteredTasks
+
+	filteredVariantTasks := []patch.VariantTasks{}
+	allFailedPlusNeededTasks := failedTaskDisplayNames
+	for _, vt := range reusePatch.VariantsTasks {
+		// limit it to tasks that are failed or who have failed tasks depending on them.
+		// we only need to add dependencies and task group tasks for failed tasks because otherwise
+		// we can rely on them being there from the previous patch.
+		if failedOnly {
+			failedPlusNeeded, err := AddTasksNeededByFailedForReuse(failedTasks, failedTaskDisplayNames, project, vt)
+			if err != nil {
+				return errors.Wrap(err, "getting dependencies for activated tasks")
+			}
+			allFailedPlusNeededTasks = append(allFailedPlusNeededTasks, failedPlusNeeded...)
+			filteredTasks = allFailedPlusNeededTasks
 		}
-		failedTasks = append(failedTasks, tasks...)
+
+		variantTask := vt
+		variantTask.Tasks = utility.StringSliceIntersection(filteredTasks, vt.Tasks)
+
+		if len(variantTask.Tasks) != 0 || len(variantTask.DisplayTasks) != 0 {
+			filteredVariantTasks = append(filteredVariantTasks, variantTask)
+			patchDoc.BuildVariants = append(patchDoc.BuildVariants, vt.Variant)
+		}
+
 	}
 
-	patchDoc.Tasks = failedTasks
+	if failedOnly {
+		patchDoc.Tasks = allFailedPlusNeededTasks
+	}
+	patchDoc.VariantsTasks = filteredVariantTasks
+
 	return nil
 }
 
+func filterToActiveForReuse(reusePatch *patch.Patch, project *model.Project) ([]task.Task, error) {
+	reuseTasks, reuseVariants := getReuseTasksAndVariants(reusePatch, project)
+	query := db.Query(bson.M{
+		task.VersionKey:      reusePatch.Version,
+		task.DisplayNameKey:  bson.M{"$in": reuseTasks},
+		task.ActivatedKey:    true,
+		task.DisplayOnlyKey:  bson.M{"$ne": true},
+		task.BuildVariantKey: bson.M{"$in": reuseVariants},
```

Comment:
Why do we need the variant/task display name parts of the query at all? Can't we just get all activated tasks? It's not clear to me what we'd be filtering out by including this.